### DISCUSSION TITLE
Hide/disable viewer window and avoid "graphics computation".

### DISCUSSION THREAD

**Discussion by DarioMaggiari** [openscenegraph/OpenSceneGraph#1234] [Q&A]
Hello,

I'm using OSG 3.4.0. 
I need to hide/disable the viewer window, mantaining the underlay computation unaltered (expecially for BB and collisions detection). 

Basically, I need to keep the scenegraph update synchronized with my data (via OSG callback mechanism already implemented) but I have to skip the rendering on viewer.  

Obviously, I think the first workaround should be to setup a traits with width and height at 0px, but I need the most correct and performance solution to keep away any unwanted overhead (because the "graphics part" is really unuseful in this situation).

Can you suggest me a way?

Many thanks,
Bye

**Comment by vsg-dev** [comment]
Have a look at the implementation of render on demand code in Viewer::run(), essentially the update and event traversals are done but nothing is rendered unless a dirty state is matched.

https://github.com/openscenegraph/OpenSceneGraph/blob/master/src/osgViewer/ViewerBase.cpp#L685

You needn't use Viewer::run() yourself, you could implement something similar in your own frameloop.

**Reply by DarioMaggiari** [reply]
Hello,

Thanks for your point. 
How to avoid to realize the window? Basically, avoiding the renderingTraversals() call in a custom frame() method seems to work as expected but the main "black frame" is displayed. I'm avoiding to call realize() method in my code (not in custom run() and frame() methods) but I'm not sure this is the correct way.

Many thanks.

**Reply by vsg-dev** [reply]
Calling Viewer::realize() does lots of crucial work to set up the OpenGL graphics context and compile GL objects, it's not surprising that things are broke if you don't call it.

You have access to the OSG code base so just have a look through the Viewer::realize() implementation to see what stuff it's doing.

**Reply by DarioMaggiari** [reply]
Basically now there are no broken parts now (in effect I see the realize() is called in some point in the code). I only need to don't show the graphics context (now it's a black screen but is ok because there is no rendering traversal as I want, I only need event and update traversal as you said) and then I'm done :D

**Comment by LaurensVoerman** [comment]
Seems to me like you have the not rendering part working, and now you want
to not show the window. Osg doesn't provide this option, so you need to go
to the underlying windowing system: for windows the code looks like:
#include <osgViewer/api/Win32/GraphicsWindowWin32>
void
viewer::setWindowVisibility(bool on)
{
 if (_graphicsWindow.valid()) {
osgViewer::GraphicsWindowWin32* win =
dynamic_cast<osgViewer::GraphicsWindowWin32*>(_graphicsWindow.get());
HWND hwin = win->getHWND();
if (hwin) {
ShowWindow(hwin, on ? SW_SHOW : SW_HIDE);
}

If you would prefer to NEVER show a window, you can change bool
GraphicsWindowWin32::realizeImplementation() (in
OpenScenGraph\src\osgViewer\GraphicsWindowWin32.cpp)
to call SetWindowPos with the "SWP_HIDEWINDOW" instead of the
"SWP_SHOWWINDOW" flag.

Regards, laurens.
